enterprise/metrics/src/main/java/org/neo4j/metrics/output/PrometheusOutput.java

```diff
@@ -0,0 +1,106 @@
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.metrics.output;
+
+import com.codahale.metrics.Counter;
+import com.codahale.metrics.Gauge;
+import com.codahale.metrics.Histogram;
+import com.codahale.metrics.Meter;
+import com.codahale.metrics.MetricRegistry;
+import com.codahale.metrics.Timer;
+import io.prometheus.client.CollectorRegistry;
+import io.prometheus.client.dropwizard.DropwizardExports;
+import io.prometheus.client.exporter.HTTPServer;
+
+import java.util.Map;
+import java.util.SortedMap;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.neo4j.helpers.HostnamePort;
+import org.neo4j.kernel.lifecycle.Lifecycle;
+import org.neo4j.logging.Log;
+
+/**
+ * Prometheus poll data from clients, this exposes a HTTP endpoint at a configurable port.
+ */
+public class PrometheusOutput implements Lifecycle, EventReporter
+{
+    private final HostnamePort hostnamePort;
+    private final MetricRegistry registry;
+    private final Log logger;
+
+    private HTTPServer server;
+    private Map<String,Object> registeredEvents = new ConcurrentHashMap<>();
+
+    public PrometheusOutput( HostnamePort hostnamePort, MetricRegistry registry, Log logger )
+    {
+        this.hostnamePort = hostnamePort;
+        this.registry = registry;
+        this.logger = logger;
+    }
+
+    @Override
+    public void init()
+    {
+        // Setup prometheus collector
+        CollectorRegistry.defaultRegistry.register( new DropwizardExports( registry ) );
+    }
+
+    @Override
+    public void start() throws Throwable
+    {
+        if ( server == null )
+        {
+            server = new HTTPServer( hostnamePort.getHost(), hostnamePort.getPort(), true );
+            logger.info( "Started publishing Prometheus metrics at http://" + hostnamePort + "/metrics" );
+        }
+    }
+
+    @Override
+    public void stop()
+    {
+        if ( server != null )
+        {
+            server.stop();
+            server = null;
+            logger.info( "Stopped Prometheus endpoint at http://" + hostnamePort + "/metrics" );
+        }
+    }
+
+    @Override
+    public void shutdown()
+    {
+        this.stop();
+    }
+
+    @Override
+    public void report( SortedMap<String,Gauge> gauges, SortedMap<String,Counter> counters,
+            SortedMap<String,Histogram> histograms, SortedMap<String,Meter> meters, SortedMap<String,Timer> timers )
+    {
+        // Prometheus does not support events, just output the latest event that occurred
+        String metricKey = gauges.firstKey();
```

Comment:
this looks strange: we already registered export of registry in the init and here again during reporting we can potentially register some other set of events? how is that possible?

Reply:
This is for the cases where we do not register, CHECK_POINT_DURATION and LOG_ROTATION_DURATION. But I did not test how this affects logging if both csv and Prometheus is activated at the same time.